File: src/odometry.py
#!/usr/bin/env python

import rospy
import numpy as np
from create_fundamentals.srv import *
from create_fundamentals.msg import SensorPacket
from green_fundamentals.msg import Position

wheel_radius = 0.036 # meters
wheelbase = 0.235  # distance between two wheels (meters)

pub = rospy.Publisher('odometry', Position, queue_size=1)
position = np.array([0, 0])
orientation = np.array([1, 0])

el_old = 0
er_old = 0

def get_wheel_distance(encoder_l, encoder_r):
    return (encoder_l - el_old)*wheel_radius, (encoder_r - er_old)*wheel_radius

def rotation_matrix(theta):
  return np.array([[np.cos(theta), -np.sin(theta)], [np.sin(theta), np.cos(theta)]])

def callback(data):
    global el_old, er_old, position, orientation

    el_new = data.encoderLeft
    er_new = data.encoderRight

    dl, dr = get_wheel_distance(el_new, er_new)

    d = (dl + dr)/2
    delta_theta = (dr - dl)/wheelbase

    orientation = rotation_matrix(delta_theta) @ orientation

    position = position + d * orientation

    msg = Position()
    msg.pos_x = position[0]
    msg.pos_y = position[1]
    msg.orientation_x = orientation[0]
    msg.orientation_y = orientation[1]
    pub.publish(msg)

    el_old = el_new
    er_old = er_new

def main():
    rospy.init_node('odometry_node', anonymous=True)
    rospy.loginfo("Starting Node...")
    rospy.wait_for_service('reset_encoders')
    reset_encoders = rospy.ServiceProxy('reset_encoders', ResetEncoders)
    reset_encoders()
    rospy.Subscriber("sensorPacket", SensorPacket, callback)
    rospy.spin()

if __name__ == '__main__':
    main()

File: src/odometry.cpp
#include "ros/ros.h"
#include <boost/bind.hpp>
#include <cmath>

#include "create_fundamentals/SensorPacket.h"
#include "green_fundamentals/Position.h"
#include "robot_constants.h"

#define EPS 0.001

float x = 0.;
float y = 0.;
float theta = 0.;

float v = 0.;
float w = 0.;

float leftEncoderOld = 0.;
float rightEncoderOld = 0.;
float timestepOld = 0.;

bool isFirstData = true;

void sensor_callback(const create_fundamentals::SensorPacket::ConstPtr& sensor_packet, ros::Publisher& odo_pub)
{ 
  if (isFirstData) {
    leftEncoderOld = msg->encoderLeft;
    rightEncoderOld = msg->encoderRight;
    timestepOld = ros::Time::now();
    firstisFirstDataData = false;
  }

  float deltaDist = 0.0f;
  float deltaX = 0.0f;
  float deltaY = 0.0f;
  float deltaTheta = 0.0f;
  float leftWheelDist = 0.0f;
  float rightWheelDist = 0.0f;
  float wheelDistDiff = 0.0f;

  float timestepNow = ros::Time::now();
  float delta_t = timestepNow - timestepOld;
  timestepOld = timestepNow;
  
  // Calculate Distance Wheels traveled
  float leftEncoderNew = msg->encoderLeft; // rad
  float rightEncoderNew = msg->encoderRight; // rad

  float leftWheelDist = (leftEncoderNew - leftEncoderOld) * WHEEL_RADIUS; // m
  float rightWheelDist = (rightEncoderNew - rightEncoderOld) * WHEEL_RADIUS; // m

  deltaDist = (rightWheelDist + leftWheelDist)/2.; // Distance robot traveled
  wheelDistDiff = rightWheelDist - leftWheelDist; 
  deltaTheta = wheelDistDiff / WHEEL_BASE;

  if (fabs(wheelDistDiff) < EPS) {
    deltaX = deltaDist * cos(theta);
    deltaY = deltaDist * sin(theta);
  } else {
    float turnRadius = (WHEEL_BASE / 2.0) * (leftWheelDist + rightWheelDist) / wheelDistDiff;
    deltaX = turnRadius * (sin(theta + deltaTheta) - sin(theta));
    deltaY = -turnRadius * (cos(theta + deltaTheta) - cos(theta));
  }

  if (fabs(delta_t) > EPS) { // hier noch einstellen was genau das ist
    v = deltaDist / delta_t;
    w = deltaTheta / delta_t;
  } else {
    v = 0;
    w = 0;
  }

  x += deltaX;
  y += deltaY;
  theta += deltaTheta;

  green_fundamentals::Position odo_msg;
  odo_msg.x = x;
  odo_msg.y = y;
  odo_msg.theta = theta;
  odo_pub.publish(odo_msg);
}


int main(int argc, char **argv)
{
  ros::init(argc, argv, "odometry");
  ros::NodeHandle n;
  ros::Publisher odo_pub = n.advertise<green_fundamentals::Position>("odometry", 1);
  ros::Subscriber sub = n.subscribe("sensor_packet", 1, boost::bind(sensor_callback, _1, odo_pub));
  ros::spin();
  return 0;
}
